File: android/src/main/java/com/livetracking/permissions/PermissionHandler.kt
package com.livetracking.permissions

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Build
import androidx.core.content.ContextCompat

/**
 * Represents the result of a permission check.
 */
sealed class PermissionResult {
    /** All required permissions are granted. */
    object Granted : PermissionResult()

    /** One or more permissions are denied or a required service is disabled. */
    data class Denied(
        val errorCode: String,
        val message: String
    ) : PermissionResult()
}

/**
 * Handles location permission checks for the LiveTracking library.
 * Provides clear error codes when permissions are denied or GPS is disabled.
 *
 * Validates: Requirements 10.1, 10.2
 */
class PermissionHandler {

    companion object {
        const val ERROR_PERMISSION_DENIED = "PERMISSION_DENIED"
        const val ERROR_GPS_DISABLED = "GPS_DISABLED"
    }

    /**
     * Checks if fine and coarse location permissions are granted.
     *
     * @param context Android context used for permission checks
     * @return [PermissionResult.Granted] if both ACCESS_FINE_LOCATION and ACCESS_COARSE_LOCATION
     *         are granted, [PermissionResult.Denied] otherwise with error code PERMISSION_DENIED
     */
    fun checkLocationPermissions(context: Context): PermissionResult {
        val fineLocationGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        val coarseLocationGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        return if (fineLocationGranted && coarseLocationGranted) {
            PermissionResult.Granted
        } else {
            val deniedPermissions = mutableListOf<String>()
            if (!fineLocationGranted) deniedPermissions.add("ACCESS_FINE_LOCATION")
            if (!coarseLocationGranted) deniedPermissions.add("ACCESS_COARSE_LOCATION")

            PermissionResult.Denied(
                errorCode = ERROR_PERMISSION_DENIED,
                message = "Location permissions denied: ${deniedPermissions.joinToString(", ")}. " +
                    "Please grant location permissions to enable tracking."
            )
        }
    }

    /**
     * Checks if background location permission is granted (Android 10+ / API 29+).
     * On devices below Android 10, background location is implicitly granted
     * when foreground location is granted.
     *
     * @param context Android context used for permission checks
     * @return [PermissionResult.Granted] if background location is available,
     *         [PermissionResult.Denied] otherwise with error code PERMISSION_DENIED
     */
    fun checkBackgroundLocationPermission(context: Context): PermissionResult {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            // Below Android 10, background location is granted with foreground location
            return PermissionResult.Granted
        }

        val backgroundLocationGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_BACKGROUND_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        return if (backgroundLocationGranted) {
            PermissionResult.Granted
        } else {
            PermissionResult.Denied(
                errorCode = ERROR_PERMISSION_DENIED,
                message = "Background location permission denied: ACCESS_BACKGROUND_LOCATION. " +
                    "Please grant 'Allow all the time' location permission for background tracking."
            )
        }
    }

    /**
     * Checks if GPS/location services are enabled on the device.
     *
     * @param context Android context used to access LocationManager
     * @return true if GPS provider is enabled, false otherwise
     */
    fun isGpsEnabled(context: Context): Boolean {
        val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as? LocationManager
            ?: return false
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
    }

    /**
     * Performs a comprehensive check of all permissions and services required for tracking.
     * Checks location permissions, background location, and GPS status in order.
     *
     * @param context Android context
     * @return [PermissionResult.Granted] if all checks pass, or the first
     *         [PermissionResult.Denied] encountered
     */
    fun checkAllTrackingRequirements(context: Context): PermissionResult {
        // Check foreground location permissions first
        val locationResult = checkLocationPermissions(context)
        if (locationResult is PermissionResult.Denied) {
            return locationResult
        }

        // Check background location permission
        val backgroundResult = checkBackgroundLocationPermission(context)
        if (backgroundResult is PermissionResult.Denied) {
            return backgroundResult
        }

        // Check GPS is enabled
        if (!isGpsEnabled(context)) {
            return PermissionResult.Denied(
                errorCode = ERROR_GPS_DISABLED,
                message = "GPS/Location services are disabled. " +
                    "Please enable location services in device settings."
            )
        }

        return PermissionResult.Granted
    }
}

File: ios/PermissionHandler.swift
import Foundation
import CoreLocation

/**
 * Represents the result of a permission or service check.
 */
enum PermissionResult {
    /// All required permissions/services are available.
    case granted
    /// A permission is denied or a required service is disabled.
    case denied(errorCode: String, message: String)
}

/**
 * Handles location permission checks for the LiveTracking library on iOS.
 * Provides clear error codes when permissions are denied or GPS is disabled.
 *
 * Validates: Requirements 10.1, 10.2
 */
class PermissionHandler {

    // MARK: - Error Code Constants

    static let ERROR_PERMISSION_DENIED = "PERMISSION_DENIED"
    static let ERROR_GPS_DISABLED = "GPS_DISABLED"

    // MARK: - Private Properties

    private let locationManager: CLLocationManager

    // MARK: - Initialization

    init(locationManager: CLLocationManager = CLLocationManager()) {
        self.locationManager = locationManager
    }

    // MARK: - Permission Checks

    /**
     * Checks the current location permission status.
     *
     * Uses `CLLocationManager.authorizationStatus()` for iOS < 14,
     * and `locationManager.authorizationStatus` for iOS 14+.
     *
     * - Returns: `.granted` if authorized (whenInUse or always),
     *            `.denied` with PERMISSION_DENIED error code otherwise.
     */
    func checkLocationPermission() -> PermissionResult {
        let status: CLAuthorizationStatus

        if #available(iOS 14.0, *) {
            status = locationManager.authorizationStatus
        } else {
            status = CLLocationManager.authorizationStatus()
        }

        switch status {
        case .authorizedWhenInUse, .authorizedAlways:
            return .granted

        case .denied:
            return .denied(
                errorCode: PermissionHandler.ERROR_PERMISSION_DENIED,
                message: "Location permission denied by user. Please grant location permission in Settings to enable tracking."
            )

        case .restricted:
            return .denied(
                errorCode: PermissionHandler.ERROR_PERMISSION_DENIED,
                message: "Location permission restricted. Location services may be restricted by parental controls or device policy."
            )

        case .notDetermined:
            return .denied(
                errorCode: PermissionHandler.ERROR_PERMISSION_DENIED,
                message: "Location permission not determined. Please request location permission before starting tracking."
            )

        @unknown default:
            return .denied(
                errorCode: PermissionHandler.ERROR_PERMISSION_DENIED,
                message: "Unknown location authorization status. Please check location permissions in Settings."
            )
        }
    }

    /**
     * Checks if location services are enabled on the device.
     *
     * - Returns: `.granted` if location services are enabled,
     *            `.denied` with GPS_DISABLED error code if disabled.
     */
    func checkLocationServicesEnabled() -> PermissionResult {
        if CLLocationManager.locationServicesEnabled() {
            return .granted
        } else {
            return .denied(
                errorCode: PermissionHandler.ERROR_GPS_DISABLED,
                message: "GPS/Location services are disabled. Please enable location services in device Settings."
            )
        }
    }

    /**
     * Performs a comprehensive check of all requirements for location tracking.
     * Checks location services enabled first, then permission status.
     *
     * - Returns: `.granted` if all checks pass, or the first `.denied` encountered.
     */
    func checkAllRequirements() -> PermissionResult {
        // Check location services enabled first
        let servicesResult = checkLocationServicesEnabled()
        if case .denied = servicesResult {
            return servicesResult
        }

        // Then check permission status
        let permissionResult = checkLocationPermission()
        if case .denied = permissionResult {
            return permissionResult
        }

        return .granted
    }

    /**
     * Requests "Always" location authorization from the user.
     * This enables background location updates.
     *
     * Note: NSLocationAlwaysAndWhenInUseUsageDescription must be configured in Info.plist.
     *
     * - Parameter locationManager: The CLLocationManager instance to request authorization on.
     */
    func requestAlwaysAuthorization(locationManager: CLLocationManager) {
        locationManager.requestAlwaysAuthorization()
    }

    /**
     * Requests "When In Use" location authorization from the user.
     *
     * Note: NSLocationWhenInUseUsageDescription must be configured in Info.plist.
     *
     * - Parameter locationManager: The CLLocationManager instance to request authorization on.
     */
    func requestWhenInUseAuthorization(locationManager: CLLocationManager) {
        locationManager.requestWhenInUseAuthorization()
    }
}
